saleor/discount/migrations/0049_migrate_sales_to_promotions.py

```diff
@@ -108,41 +137,79 @@ def migrate_translations(
+    CheckoutLineDiscount, sales_pks, rule_by_channel_and_sale
 ):
     if checkout_line_discounts := CheckoutLineDiscount.objects.filter(
         sale_id__in=sales_pks
     ).select_related("line__checkout"):
         for checkout_line_discount in checkout_line_discounts:
-            channel_id = checkout_line_discount.line.checkout.channel_id
-            sale_id = checkout_line_discount.sale_id
-            lookup = f"{channel_id}_{sale_id}"
-            if promotion_rule := rule_by_channel_and_sale.get(lookup):
-                checkout_line_discount.promotion_rule = promotion_rule
+            if checkout_line := checkout_line_discount.line:
+                channel_id = checkout_line.checkout.channel_id
+                sale_id = checkout_line_discount.sale_id
+                lookup = f"{channel_id}_{sale_id}"
+                if promotion_rule := rule_by_channel_and_sale.get(lookup):
+                    checkout_line_discount.promotion_rule = promotion_rule
 
         CheckoutLineDiscount.objects.bulk_update(
             checkout_line_discounts, ["promotion_rule_id"]
         )
 
 
 def migrate_order_line_discounts(
-    sales_pks, OrderLineDiscount, rule_by_channel_and_sale
+    OrderLineDiscount, sales_pks, rule_by_channel_and_sale
 ):
     if order_line_discounts := OrderLineDiscount.objects.filter(
         sale_id__in=sales_pks
     ).select_related("line__order"):
         for order_line_discount in order_line_discounts:
-            channel_id = order_line_discount.line.order.channel_id
-            sale_id = order_line_discount.sale_id
-            lookup = f"{channel_id}_{sale_id}"
-            if promotion_rule := rule_by_channel_and_sale.get(lookup):
-                order_line_discount.promotion_rule = promotion_rule
+            if order_line := order_line_discount.line:
+                channel_id = order_line.order.channel_id
+                sale_id = order_line_discount.sale_id
+                lookup = f"{channel_id}_{sale_id}"
+                if promotion_rule := rule_by_channel_and_sale.get(lookup):
+                    order_line_discount.promotion_rule = promotion_rule
 
         OrderLineDiscount.objects.bulk_update(
             order_line_discounts, ["promotion_rule_id"]
         )
 
 
+def get_rule_by_channel_sale(rules_info):
+    return {
+        f"{rule_info.channel_id}_{rule_info.sale_id}": rule_info.rule
+        for rule_info in rules_info
+    }
+
+
+def channel_listing_in_batches(qs):
+    first_sale_id = 0
+    while True:
+        batch_1 = qs.filter(sale_id__gt=first_sale_id)[:BATCH_SIZE]
+        if len(batch_1) == 0:
+            break
+        last_sale_id = batch_1[len(batch_1) - 1].sale_id
+
+        # `batch_2` extends initial `batch_1` to include all records from
+        # `SaleChannelListing` which refer to `last_sale_id`
+        batch_2 = qs.filter(sale_id__gt=first_sale_id, sale_id__lte=last_sale_id)
+        pks = list(batch_2.values_list("pk", flat=True))
```

Comment:
Almost the same as above, just fetch `pk` and `sale_id`, it will also reduce hits to database.